File: 2024/go/02/part1.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)

func main() {
	file, err := os.Open("input.dat")
	if err != nil {
		fmt.Println("Error opening file:", err)
		return
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)

	cnt := 0
	for scanner.Scan() {
		parts := strings.Fields(scanner.Text())

		safe := isSafe(parts)
		if safe {
			cnt++
		}
	}
	if err := scanner.Err(); err != nil {
		fmt.Println("Error reading file:", err)
		return
	}

	fmt.Println(cnt)
}

func isSafe(report []string) bool {
	monotoneus := ""
	for i := 1; i < len(report); i++ {
		num1, err := strconv.Atoi(report[i-1])
		if err != nil {
			fmt.Println("Error converting number:", err)
			return false
		}

		num2, err := strconv.Atoi(report[i])
		if err != nil {
			fmt.Println("Error converting number:", err)
			return false
		}

		if (int(math.Abs(float64(num1-num2))) == 0 || int(math.Abs(float64(num1-num2))) > 3) ||
			(monotoneus == "asc" && num1 > num2) ||
			(monotoneus == "desc" && num1 < num2) {
			return false
		}

		if monotoneus == "" {
			if num1 < num2 {
				monotoneus = "asc"
			} else {
				monotoneus = "desc"
			}
		}
	}

	return true
}

File: 2024/python/02/part1.py
# Read the input.dat each line has to values one for the left list and one for the right list
raw = open('input.dat').readlines()

cnt = 0
for report in raw:
    levels = [int(x) for x in report.split()]
    
    safe = True
    monotoneus = ''
    for i in range(1, len(levels)):
        if abs(levels[i] - levels[i  - 1]) == 0 or abs(levels[i] - levels[i  - 1]) > 3 \
            or monotoneus == 'des' and levels[i  - 1] < levels[i] \
            or monotoneus == 'asc' and levels[i  - 1] > levels[i]:
            safe = False
            break
        else:
            monotoneus = 'asc' if levels[i  - 1] < levels[i] else 'des'
    
    if safe:
        cnt += 1

print(cnt)
